code/iaas/ha/src/main/java/io/cattle/platform/ha/monitor/impl/PingInstancesMonitorImpl.java

```diff
@@ -56,12 +90,18 @@ public void pingReply(Ping ping) {
         }
 
         long agentId = Long.parseLong(ping.getResourceId());
-        Map<String, String> knownInstances = instanceCache.getUnchecked(agentId);
+        Map<String, Instance> knownInstances = instanceCache.getUnchecked(agentId);
+
+        Agent agent = objectManager.loadResource(Agent.class, agentId);
+        Map<String, Host> hosts = agentDao.getHosts(agent.getId());
+        Host host = hosts.get(reportedInstances.hostUuid);
 
-        if (handleUnreportedKnownInstances(agentId, knownInstances, reportedInstances, true)) {
+        try {
+            syncContainers(knownInstances, reportedInstances, agent.getAccountId(), agent.getId(), host.getId(), true);
+        } catch (ContainersOutOfSync e) {
             knownInstances = load(agentId);
             instanceCache.put(agentId, knownInstances);
-            handleUnreportedKnownInstances(agentId, knownInstances, reportedInstances, false);
+            syncContainers(knownInstances, reportedInstances, agent.getAccountId(), agent.getId(), host.getId(), false);
```

Comment:
This question is more for my understanding, but ultimately, the state of the host or rancher can change after the ping data or after we refresh the knownInstances in the cache.  So the syncActions that we generate can potentially be out of date by the time they're executed.  I noticed in the python-agent changes, we handle this already.  Is this correct?